### PR TITLE
KAFKA-5461: Add metric to track global topic count and global parition count in a cluster

### PR DESCRIPTION
**Pull request by abhishekmendhekar**

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6171/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6155/
Test PASSed (JDK 8 and Scala 2.12).

File: core/src/main/scala/kafka/controller/KafkaController.scala [hunk 1]
```diff
@@ -1532,6 +1550,14 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, time: Time, met
             !topicDeletionManager.isTopicQueuedUpForDeletion(topicPartition.topic)
         }
       }
+
+    globalTopicCount =
+      if (!isActive) 0
+      else controllerContext.allTopics.size
```
Comment (jjkoshy) [inline core/src/main/scala/kafka/controller/KafkaController.scala]: Can you make these confirm to the existing indentation conventions in use?

**Review by jjkoshy — COMMENTED** [review]
Can you look into possibly enhancing `MetricsTest` to check for correctness of these metrics?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6175/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6159/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by omkreddy** [comment]
LGTM

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, code changes look good, just need some simple tests like @jjkoshy suggested.

File: core/src/main/scala/kafka/controller/KafkaController.scala (hunk 1 above)
Comment (abhishekmendhekar) [reply core/src/main/scala/kafka/controller/KafkaController.scala]: @jjkoshy Made it consistent with other if-else blocks.

**Comment by abhishekmendhekar** [comment]
@jjkoshy @ijuma @omkreddy Thank you all for the review! The new patch has tests to validate the new metrics.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6223/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6207/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by abhishekmendhekar** [comment]
The test setup starts with two servers. So the controller could be either Broker 0 or Broker 1 but the `DefaultMetricRegistry` is a single static instance. There is a race on who updates the metrics last and reflect those values. These test failure will be random.

Possible two solutions: Changes the number of brokers in MetricTest to 1 or add a test to KafkaControllerTest. I will try the first solution first and if that doesnt work then the second one.

**Comment by abhishekmendhekar** [comment]
@jjkoshy @ijuma Looks like the static nature of the `DefaultMetricRegistry` make unit testing really hard. I guess it isn't as straight forward to unit-test controller metrics. Based on the actual code, I feel testing metrics is an over-kill. IMO, we should proceed without adding tests. Let me know if this works.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6227/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6211/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by jjkoshy** [comment]
Yes I think you are right wrt the issue with the way we use yammer metrics.
+1 on the patch, but can you fix the indentation on the `if` statement?
```
if (!isActive) {
  0
} else {
  controllerContext.partitionLeadershipInfo.size
}
```

**Comment by ijuma** [comment]
Hmm, we don't require braces for Scala if/else statements. About tests, we do tend to write simple tests that just check that the metric is present and the value is within the acceptable range. Can't that be done here?

**Comment by jjkoshy** [comment]
I'm going with samples from `git grep "= if ("` - we aren't really consistent, but I was actually referring to the spacing of the else.
WRT the tests: not sure if existence is useful and the range check may not work given the clobbering of metrics, unless I misunderstood Abhishek's observation.

**Comment by ijuma** [comment]
Given that the metrics are derived from the class name, it's actually useful to have tests as you can break metrics by simply renaming the class containing them. Or moving them to an inner class accidentally. Code with no tests is one refactoring away from being broken IMO.

**Comment by abhishekmendhekar** [comment]
@ijuma I agree with adding tests and that is exactly what I started with (which even tested purging of metrics on failover). But the problem is the way in which the tests are written today for metrics. The metric registry object being static may get clobbered but doesn't mean all the metrics get clobbered. So there may be tests that run perfectly fine but we can't guarantee the correctness (as and when new tests get added). The new metrics I added exposes this issue very clearly. The solution for this is not to use static object but using metric registry per broker instance but this might be a new KIP altogether. I would say that having a new KIP to get rid of static registry object is the right solution instead of adding unreliable tests.

@jjkoshy You are right about the if/else indentations being inconsistent. We must investigate and add a style checker to enforce certain coding practices.

**Comment by abhishekmendhekar** [comment]
@jjkoshy [Here](https://github.com/apache/kafka/pull/3549/files#diff-ed90e8ecc5439a5ede5e362255d11be1R1534) are the indentations I followed. Let me if this looks good. Thanks!

**Comment by abhishekmendhekar** [comment]
Ping @ijuma @jjkoshy! 😄

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6415/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6400/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by jjkoshy** [comment]
I pinged you offline with the indentation style - yes we should have a style checker for these details. See KAFKA-2423 for e.g.

I think Ismael's point on an existence check to at least catch inadvertent naming changes is valid given that metrics are effectively public API.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6464/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by abhishekmendhekar** [comment]
For some reason, Jenkins didn't pick the changes from github to run the tests. Restarting tests.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6482/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6467/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6501/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6486/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6491/
Test PASSed (JDK 8 and Scala 2.12).

**Review by ijuma — APPROVED** [review]
Thanks for adding the test. LGTM apart from the unnecessary braces. The Scala style guide (http://docs.scala-lang.org/style/control-structures.html#curly-braces) says to omit braces if the each line of the if/else is a single expression. In the interest of making progress, I'll fix that before the merge.

Merging to trunk.

**Pull request closed without merging**